### ISSUE TITLE
SD Card Mount Failed

### ISSUE THREAD

**Issue by joypictures** [tiredboffin/sdwifi#4]
Hi, thank you very much for you great project, so now my sdwifi pro can works! But I'm experiencing a strange issue, maybe an hardware fault or an ESP32 configuration. After some time if I stop work on the SD with http calls, the card goes into a state where it can't access the memory, the sd filesystem. Firstly I noticed it with sd into an EOS Canon camera and I've thought on a power fault from the camera, but now I've the sd in the programmer and I can log what happens:
[59642773][I][sdwifi.ino:272] mountSD(): SD Card mount
E (59642827) vfs_fat_sdmmc: mount_to_vfs failed (0x101).
[59642831][E][SD_MMC.cpp:190] begin(): Failed to initialize the card (0x101). Make sure SD card lines have pull-up resistors in place.
[59642848][E][sdwifi.ino:284] mountSD(): SD Card Mount Failed

Obviously if I power off and on again all restarts well.
ciao, Davide

**Comment by dbucci69** [comment]
I wrote with company account, this is mine (Davide), I've a windows app that polls every 5 seconds the cards with list command.  I can reproduce the issue easily. Power on the card and wait some hours. It enters in this state where it can't access the sd filesystem. While I was here writing the log started to write this:
[60157936][I][sdwifi.ino:272] mountSD(): SD Card mount
[60158001][W][WebServer.cpp:572] send(): content length is zero
[60158205][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[60158864][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[60159866][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[60160866][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[60161866][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[60162866][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"

Hope this can help.

Davide

**Comment by tiredboffin** [comment]
I’m a bit surprised it handles that many requests 😄 In my workflow (testing camera firmware), the camera reboots during every testing cycle, which also reboots the ESP32. So, it usually doesn’t get a chance to process more than 10–20 requests before restarting.  

That said, it’d be interesting to fix this, thank you for the report! However I’m still not entirely clear on how to reproduce the issue. Here’s what I’ve understood so far:  

1) The ESP32 eventually gets stuck and can’t mount the SD card.  
2) The web server keeps running but starts returning errors for commands that interact with the SD card.  
3) This happens even if the card is in the development board and isn’t connected to a host (like a camera or PC).  
4) To reproduce it, I’d need to poll the web server with the "list" command for N hours until it starts returning errors.  

Does that sound about right?  

In the meantime, could you try these:  

a) Run the "power=reset" command and see if it reboots and resolves the issue. If point (2) is correct, this might be a quick workaround.  
b) Increase the polling frequency in your setup to 2Hz (from 0.2Hz) and see if that affects how long it takes for the issue to appear.  

Thanks,
tb

**Comment by dbucci69** [comment]
I speeded up the poll and get the error early. I've experienced more things:

**Comment by dbucci69** [comment]
I speeded up the poll and get the error early. I've experienced more things:
there's different errors, and different webserver repsonse.
```
[13501252][I][sdwifi.ino:272] mountSD(): SD Card mount
[13501314][W][WebServer.cpp:572] send(): content length is zero
[13501338][I][sdwifi.ino:303] umountSD(): In SD Card Unmount
[13501391][I][sdwifi.ino:272] mountSD(): SD Card mount
[13501456][W][WebServer.cpp:572] send(): content length is zero
[13502515][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13502592][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13502888][I][sdwifi.ino:303] umountSD(): In SD Card Unmount
[13503199][I][sdwifi.ino:272] mountSD(): SD Card mount
[13503264][W][WebServer.cpp:572] send(): content length is zero
[13503287][I][sdwifi.ino:303] umountSD(): In SD Card Unmount
[13504323][I][sdwifi.ino:272] mountSD(): SD Card mount
[13504385][W][WebServer.cpp:572] send(): content length is zero
[13504410][I][sdwifi.ino:303] umountSD(): In SD Card Unmount
[13504438][I][sdwifi.ino:272] mountSD(): SD Card mount
[13504503][W][WebServer.cpp:572] send(): content length is zero
[13504709][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13505174][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13505642][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13506072][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13506498][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13506925][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13507151][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13507526][I][sdwifi.ino:303] umountSD(): In SD Card Unmount
[13507909][I][sdwifi.ino:272] mountSD(): SD Card mount
[13507974][W][WebServer.cpp:572] send(): content length is zero
[13507995][I][sdwifi.ino:303] umountSD(): In SD Card Unmount
[13508314][I][sdwifi.ino:272] mountSD(): SD Card mount
[13508376][W][WebServer.cpp:572] send(): content length is zero
[13508400][I][sdwifi.ino:303] umountSD(): In SD Card Unmount
[13508427][I][sdwifi.ino:272] mountSD(): SD Card mount
[13508493][W][WebServer.cpp:572] send(): content length is zero
[13508698][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13509160][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13509713][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13510717][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13511717][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13512717][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13513717][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13514719][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13515275][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"

```
when the error is errno 11, "No more processes" the webserver doesn't answer, after some minutes it comes back and answer with "LIST:SDBUSY"

```
13646744][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13647744][E][NetworkClient.cpp:429] write(): fail on fd 49, errno: 11, "No more processes"
[13647752][I][sdwifi.ino:303] umountSD(): In SD Card Unmount
[13648758][I][sdwifi.ino:272] mountSD(): SD Card mount
E (13648811) vfs_fat_sdmmc: mount_to_vfs failed (0x101).
[13648816][E][SD_MMC.cpp:190] begin(): Failed to initialize the card (0x101). Make sure SD card lines have pull-up resistors in place.
[13648832][E][sdwifi.ino:284] mountSD(): SD Card Mount Failed
```

Now I've executed, as suggested, power=reset, but the command seems to fail, this is the log from two sends:
```
[14604515][E][vfs_api.cpp:23] open(): File system is not mounted
[14604521][E][vfs_api.cpp:23] open(): File system is not mounted
[14604529][E][vfs_api.cpp:23] open(): File system is not mounted
[14604536][E][WebServer.cpp:798] _handleRequest(): request handler failed to handle request
[14604547][W][WebServer.cpp:572] send(): content length is zero
[14627859][E][vfs_api.cpp:23] open(): File system is not mounted
[14627865][E][vfs_api.cpp:23] open(): File system is not mounted
[14627873][E][vfs_api.cpp:23] open(): File system is not mounted
[14627881][E][WebServer.cpp:798] _handleRequest(): request handler failed to handle request
[14627891][W][WebServer.cpp:572] send(): content length is zero
```

webserver seems to works correctly, sysinfo responds correctly but can't access the sd:
```
[15380301][I][sdwifi.ino:272] mountSD(): SD Card mount
E (15380355) vfs_fat_sdmmc: mount_to_vfs failed (0x101).
[15380359][E][SD_MMC.cpp:190] begin(): Failed to initialize the card (0x101). Make sure SD card lines have pull-up resistors in place.
[15380376][E][sdwifi.ino:284] mountSD(): SD Card Mount Failed
```

the response is:
`{"info":{"filesystem":{"status":"failed"},.....`

**Comment by dbucci69** [comment]
I badly used the "power=reset" command, after looking in the code I've seen the correct way to invoke it, with "exp?", now it works, the esp reboot and after a while webserver starts again and sd is accessible.
Is there a way to know if the sd is really busy 'cause the camera takes control of it? I think that resetting it after some tries from my polling software or in the firmware is not the right way. If we know that the host is not using the sd and we can't mount the sd, that way we can try to reset the esp.

**Comment by tiredboffin** [comment]
> Is there a way to know if the sd is really busy 'cause the camera takes control of it? 

In a specific case, when the host uses SD in SPI mode, it’s possible to determine the card usage (mounted) status. However, in 4-bit mode , this I think can only be inferred — and even then, only if the host logic is known and predictable — by inspecting the "ISR" section of the returned sysinfo, which provides a timestamp of the most recent data transfer. 

A general solution to prevent collisions would involve controlling the host system to ensure the SD card is not in use at the same time as the ESP32. This could be done obviously by unmounting the card or stopping any processes on the host that are reading from or writing to it.

> I think that resetting it after some tries from my polling software or in the firmware is not the right way. 

Of course, this is not a solution, just a temporary workaround until a proper solution is identified and implemented.

> If we know that the host is not using the sd and we can't mount the sd, that way we can try to reset the esp.

In fact, I believe you can safely reboot the ESP32 at any time without interrupting the host's operation. This can easily be validated experimentally by running a long transfer on the host and then resetting the ESP32.

I'll try to reproduce the issue later this evening and let you know what I find.

**Comment by tiredboffin** [comment]
I was able to reproduce the issue after about an hour of sysinfo() calls at 4 requests per second. I have a working hypothesis for the cause and will validate it over the weekend.

**Comment by tiredboffin** [comment]
The root cause of this issue is a memory bug in IDF-ESP32 tools v5.1, which is used by the latest esp32:esp32@3.0.7. 

Each call to `unmount` leaks approximately 28 bytes, which can quickly accumulate and fragment the memory after around 4-5k calls, eventually causing `mount()` and potentially other functions to fail. 

This bug was [[fixed](https://github.com/espressif/esp-idf/commit/91eada4accdef58f636d4dce15e2803a7a42083f)](https://github.com/espressif/esp-idf/commit/91eada4accdef58f636d4dce15e2803a7a42083f) in v5.2. 

I’ve patched my local copy of v5.1 and confirmed that the issue has been resolved in my tests. However, I'm unsure when Arduino will release the update. 

You can try the latest release candidate at https://github.com/espressif/arduino-esp32/releases/tag/3.1.0-RC3 that is based on IDF 5.3.

In the meantime I’ve added the following:

1) **Memory statistics**: This should make it easier to identify any memory leakage.

2) **Automatic reboot**: This triggers when memory becomes too low. The logic is quite basic, but it should help with recovery.

3) **Experimental option**: An additional option to the `sysinfo?sd=none` command to bypass the calls to the leaking `mount()/unmount()` functions.

From what I understand, you’re polling the card for new files. If that’s correct, you can use the leak-free `sysinfo?sd=none` for polling, and monitor `info.isr.count` and `info.isr.count.activity_millis_ago` in the response. By comparing the values between two subsequent `sysinfo?sd=none` calls, it should be possible to detect whether the host has accessed the card. Once this is determined, you can issue the leaky `list` command to get the update. While this still requires testing, if it works, it should help reduce both the memory leakage and the wear on the SD card.

Please try out the `4-sd-card-mount-failed` branch and let me know how it goes.

Thanks, 
tb

**Comment by dbucci69** [comment]
I'm testing your branch and will let you know better: sometimes it seems to freeze just for a second, in the monitor it stops messaging, no error is reported, then restart normally, the mount conter continues to increase. meanwhile I have also patched the lib and tested: it seems to solve the issue but I get the info.filesystem.usedbytes always the same number and extremely high! different from the ones I read with other builds. Have you noted the same?

D

**Comment by dbucci69** [comment]
build based on 4-sd-card-mount-failed
I've implemented in my app the isr reading and test with previous, so when it changes I make a list call. I leave the sd in the camera and the app to poll it for the whole weekend and on Monday I find out if it is still working well!
D

**Comment by dbucci69** [comment]
hi tb, the weekend is over and this morning my app is still querying the sd card which is alive! I took some photos as soon as possible and they were transferred. I can confirm that the problem is indeed in the mount call as you discovered and asked to investigate. currently my app uses the call to sysinifo?sd=none before querying with list?path=... It works fine and can stay that way, as for the patch to the library I tried to apply it to version 5.1.4 which seems to be the one I'm using, but then when it reads the information for the sd it gets the value of usedBytes wrong. I modified this file:
./esp/esp-idf/components/fatfs/vfs/vfs_fat_sdmmc.c
and compiling an example for sdcards I replaced this library
/root/.arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-33fbade6/esp32/lib/libfatfs.a
It works fine but reports the wrong value of usedBytes, and maybe it hides other problems that I don't see now. If you have any suggestions on how to apply the patch well, I'll try again. thanks

**Comment by tiredboffin** [comment]
Thank you for the feedback! I have updated the main branch to include the sd=none workaround. The newer  arduino-esp32 candidate [RC3](https://github.com/espressif/arduino-esp32/releases/tag/3.1.0-RC3)  was released two weeks ago, so hopefully the final version should be coming soon, and it should include the fix for the memory leak bug.  However, even then, it may still make sense to continue using the sd=none option for polling to reduce the wear and tear on the SD card.
